### PR TITLE
emoji lingo app: displays a random emoji in any supported language

### PR DESCRIPTION
**Pull request by cedricsam**
TO-DO / BUGS:
* catch error when text is missing for certain emojis
* refactor and lint code
* validate/test supported languages
* make sure defaults are understood

### CODE REVIEW

**Comment by cedricsam** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by cedricsam** [comment]
Not quite ready for prime time, but would like some feedback!

File: apps/emojilingo/emojilingo.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+var source []byte
+
+// New creates a new instance of the Emoji Lingo applet.
+func New() manifest.Manifest {
+	return manifest.Manifest{
+		ID:          "emoji-lingo",
+		Name:        "Emoji Lingo",
+		Author:      "Cedric Sam",
+		Summary:     "A random emoji, localized",
+		Desc:        "Displays a random emoji and its unique short text annotation from the Unicode Consortium in a given locale.",
```
Comment (matslina) [inline apps/emojilingo/emojilingo.go:20]: Many non-technical users will likely not understand what "locale" means in this context.

File: apps/emojilingo/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Emoji Lingo for Tidbyt
+
+By: Cedric Sam
+
+See: [https://github.com/cedricsam/emoji-lingo](https://github.com/cedricsam/emoji-lingo)
```
[the remark this answers is not in the document]
Comment (cedricsam) [reply apps/emojilingo/README.md:5]: ya, i intend to publish when/if this publishes. Includes the scripts for generating new locales, scraping emojis, etc.

File: apps/emojilingo/emoji_lingo.star [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    # Pick an emoji at random (hopefully a good one)
+    number_valid_emojis = len(valid_emoji_data)
+    rand_index = random.number(0, number_valid_emojis)
+    print("Picking from %d random emojis... random index was %d" % (number_valid_emojis, rand_index))
+    name_item = valid_emoji_data[rand_index]
+
+    # Get the base64 text file
+    base64_url = EMOJI_BASE64_URL % normalizeCode(name_item[0])
+    print("Making request for emoji base64 to %s" % base64_url)
+    rep_base64 = http.get(base64_url)
```
[the remark this answers is not in the document]
Comment (cedricsam) [reply apps/emojilingo/emoji_lingo.star:78]: cool, and it's nice to know. Would love to hear more (on DMs -- no rush)

**Comment by cedricsam** [comment]
I cached the random emoji by locale (ttl = 30 seconds) because I can’t guarantee a cached emoji in one language would exist in another (TBD later).

Could you help me validate this: I calculated that the worst-case for the cached emojis was going to be 3.2GB per month, well below the S3 free tier for data transfer: 2000 bytes per emoji base64 in average * 20 locales * 2 caching events per minute * 1440 minutes per day * 30 days = 3456000000 bytes.

I cached the emoji names for 1 day (86400 seconds), since I doubt they'll change once they're up. Can it be a longer cache? Does it risk being stuck, or a different schema version busts the cache?

Thanks! (I'll make this a mergeable PR! And I'll make the GH link public when/if we near publication)

**Comment by cedricsam** [comment]
Added support for emojis from different vendors (but I haven't generated the Google ones, planning to do so sometime between now and weekend)

**Comment by cedricsam** [comment]
Added Google emoji base64 text files to the s3 bucket

**Comment by cedricsam** [comment]
bunch of bug fixes... i used an incorrect variable name w/o checking :( and then left a `print()` of my cache for testing... Tested on pixlet and it seems to work (and the data files appear to be all accounted for, including google-style base64 images)

**Comment by cedricsam** [comment]
on second thought, decided to make cache NOT taking into account language. It seems a better experience if the user can stay with a single emoji while browsing languages, as a way to appreciate that the app shows the same emojis (within caching period) in different languages...

**Comment by cedricsam** [comment]
added support for gzip decompressing of csv data files

test fails on missing gzip module, but would work on pixlet 0.17.4 (where gzip is now supported)

**Review by betterengineering — APPROVED** [review]

**Pull request merged**